src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -434,15 +434,33 @@ public void EnsureSolutionIsLoaded()
             });
         }
 
-        public string SolutionDirectory => NuGetUIThreadHelper.JoinableTaskFactory.Run(GetSolutionDirectoryAsync);
+        public string SolutionDirectory
+        {
+            get
+            {
+                if (!string.IsNullOrEmpty(_solutionDirectory))
+                {
+                    return _solutionDirectory;
+                }
+
+                _solutionDirectory = NuGetUIThreadHelper.JoinableTaskFactory.Run(GetSolutionDirectoryAsync);
+                return _solutionDirectory;
```

Comment:
I think this has a bug if a solution with only legacy csproj is opened, something calls `ISolutionManager.SolutionDirectory`, without `InitializeAsync` ever being run, then solution is closed. In this case, `OnAfterClosing` will never be called, since `InitializeAsync` was never called, and then `VsSolutionManager` will still return a solution directory for a solution that's now closed.

Understanding in what scenarios` .SolutionDirectory` is called, without `InitializeAsync()` being called is difficult. Opening PM UI, PMC, or doing a restore (including a build) will cause `VsSolutionManager.InitializeAsync()` to be called, so we have to think of different scenarios.

One example might be right clicking the solution in solution explorer, as the code that checks if the "restore nuget packages" command should be enabled/disabled calls something on `ISolutionManager`, although it's more likely to call `.IsSolutionOpen`, rather than `.SolutionDirectory`.  The analysis of checking every single reference of `.SolutionDirectory` to ensure that all of them cause `InitializeAsync()` to be called was more effort than just implementing code that should "do the right thing", however, that one failing test prevented me from merging my PR.

Reply:
> I think this has a bug if a solution with only legacy csproj is opened, something calls `ISolutionManager.SolutionDirectory`, without `InitializeAsync` ever being run, then solution is closed. In this case, `OnAfterClosing` will never be called, since `InitializeAsync` was never called, and then `VsSolutionManager` will still return a solution directory for a solution that's now closed.

I just tested with solution with only PC and PR legacy csproj, both cases `OnAfterClosing` was called and `solution directory` value reset to null. Am I missing something?

Reply:
@zivkan 
Could you be able to test possible breaking scenarios with vsix artifacts from [here](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=7024573&view=artifacts&pathAsName=false&type=publishedArtifacts) when you have time, please?
Currently it's passing RPS tests.

Reply:
I found a scenario that doesn't seem to work. 

1. Load a C++ project (make sure no PMC, or anything like that is loaded). 
2. Open the settings via `Options -> NuGet Pakcage Manager Settings`.
3. Close the settings dialog
4. Close the solution (not VS).
5. Load any other solution.

![image](https://user-images.githubusercontent.com/2878341/206334517-0f5bba0b-125e-471e-84c5-efb6203239c0.png)


This is basically what @zivkan is referring to by `ISolutionManager.SolutionDirectory, without InitializeAsync ever being run, then solution is closed`. 

In cases that the solution manager does not get initialized, we don't get a call to reset the property.

Reply:
btw you might want to rebase cause there's been so many project fixes, and it pretty much destroyed all the local caches :D

Reply:
I believe I fixed my other PR: https://github.com/NuGet/NuGet.Client/pull/4879/commits/fd029a7386b50328354b64c2e9d603712aa542ed The test passes locally, and I'm now waiting for CI results.

Therefore, I believe we can close this PR and accept mine: https://github.com/NuGet/NuGet.Client/pull/4879